### PR TITLE
feat(tactic/interactive): generalize custom tactic monads in `begin ... end` blocks

### PR DESCRIPTION
**Pull request by khoek**
In core lean in order to define a custom interaction monad (i.e. the data required to be able to use `begin ... end` and `by ...` blocks), you need to define several functions (as detailed in the header of [src/frontends/lean/tactic_notation.cpp](https://github.com/leanprover-community/lean/blob/master/src/frontends/lean/tactic_notation.cpp)). Of note are the lines:
````
7) There is a definition Tac.execute (tac : Tac unit) : tactic unit
8) There is a definition Tac.execute_with (cfg : config) (tac : Tac unit) : tactic unit
````
We generalise these pair of functions into an instance which can be made available for any custom `Tac` you would like, without hard-coding functions with these specific hardcoded names (which cannot then be overridden, for example). We instead just offload this task to the typeclass resolution mechanism. I think there is a reasonable argument that the other auxiliary functions for a custom tactic monad should just be moved into a single instance of `interactive_structure` (or some other name) for that purpose, but I leaned on the conservative side here (since that would be a simple refactoring given the new instance I've introduced anyway). As a side-effect, there is no more special casing for handling `begin ... end` tactic blocks compared to any other kind of custom interactive block.

### CODE REVIEW

File: library/init/meta/interactive_base.lean [hunk 1]
```diff
@@ -195,6 +195,33 @@ meta structure inductive_decl :=
...
+    ```
+    begin [custom_monad] with config,
+        ...
+    end
+    ```
+-/
+meta class executor (m : Type → Type u) [monad m] :=
+(execute : m unit → tactic unit)
+(config_type : Type := unit)
+(execute_with : config_type → m unit → tactic unit := λ _, execute)
```
Comment (digama0) [inline library/init/meta/interactive_base.lean]: How about having only `execute_with`, and provide a default `config_type` object to pass? That seems consistent with the smt example, and with most use cases I can think of.

Comment (khoek) [reply library/init/meta/interactive_base.lean]: Sounds great. Done.

**Comment by khoek** [comment]
Of course this should be non-breaking with mainline `lean`, and the tests + current mathlib building against this version testifies to that.

File: library/init/meta/tactic.lean [hunk 2]
```diff
@@ -73,6 +73,45 @@ meta def {u₁ u₂} tactic.down {α : Type u₂} (t : tactic (ulift.{u₁} α))
...
+    executed. The `inhabited` first argument facilitates the passing of an
+    optional configuration parameter `config`, using the syntax:
+    ```
+    begin [custom_monad] with config,
+        ...
+    end
+    ```
+-/
+meta class executor (m : Type → Type u) [monad m] :=
+{config_type : Type}
```
Comment (digama0) [inline library/init/meta/tactic.lean]: You shouldn't have `config_type` be implicit here. Conversely, does the implicitness of the `inhabited` instance allow it to be inferred in the `smt_config` example?

Comment (khoek) [reply library/init/meta/tactic.lean]: @digama0 I'm pretty sure there's no `inhabited unit` in core?

Comment (khoek) [reply library/init/meta/tactic.lean]: Oh, that has nothing to do with what you said, gotcha

Comment (digama0) [reply library/init/meta/tactic.lean]: If there is no `inhabited unit` in core, then I know what the next PR is...

Comment (khoek) [reply library/init/meta/tactic.lean]: Changed, but the instance can't be inferred in that example.

Comment (khoek) [reply library/init/meta/tactic.lean]: @digama0 I thought about that, but there is a copy in mathlib (if I recall). So we run into a problem where we wouldn't be compatible between lean versions.

**Comment by khoek** [comment]
(Travis died, let's try that one again)

**Comment by cipher1024** [comment]
@digama0 Are we happy with this PR?

File: src/library/constants.cpp [hunk 3]
```diff
@@ -158,6 +158,7 @@ name const * g_int_neg_ne_of_pos = nullptr;
 name const * g_int_ne_neg_of_pos = nullptr;
 name const * g_int_neg_ne_zero_of_ne = nullptr;
 name const * g_int_zero_ne_neg_of_ne = nullptr;
+name const * g_interactive_executor = nullptr;
```
Comment (gebner) [inline src/library/constants.cpp:161]: @cipher1024 @khoek The `constants.*` files should not be edited by hand, they are automatically generated from the `constants.txt` file using the `scripts/gen_constants_cpp.py` script.

Comment (khoek) [reply src/library/constants.cpp:161]: Oh, ok, thanks for the heads up. When I wrote this I was totally on my own in the codebase. :)

Comment (gebner) [reply src/library/constants.cpp:161]: No hard feelings.  I was just surprised when I added something to `constants.txt` so I felt like I should share this.

**Pull request merged**